src/platform/windows/display_vram.cpp

```diff
@@ -910,97 +920,228 @@ namespace platf::dxgi {
+    auto get_locked_platform_img = [&](std::shared_ptr<platf::img_t> &img) -> std::tuple<std::shared_ptr<img_d3d_t>, texture_lock_helper> {
+      auto platform_img = std::static_pointer_cast<img_d3d_t>(img);
+      if (complete_img(platform_img.get(), false)) return { nullptr, nullptr };
+      texture_lock_helper lock_helper(platform_img->capture_mutex.get());
+      if (!lock_helper.lock()) {
+        BOOST_LOG(error) << "Failed to acquire capture mutex";
+        return { nullptr, nullptr };
+      }
+      return { platform_img, std::move(lock_helper) };
+    };
 
-      status = img->capture_mutex->AcquireSync(0, INFINITE);
-      if (status != S_OK) {
-        BOOST_LOG(error) << "Failed to acquire capture mutex [0x"sv << util::hex(status).to_string_view() << ']';
-        return capture_e::error;
+    switch (last_frame_action) {
+      case lfa::nothing: {
+        break;
       }
 
-      // Populate the image with dummy data. This is required because these images could be reused
-      // after rendering (in which case they would have a cursor already rendered into them).
-      device_ctx->UpdateSubresource(img->capture_texture.get(), 0, nullptr, dummy_data.get(), img->row_pitch, 0);
-    }
-    else {
-      // We must know the capture format in this path or we would have hit the above unknown format case
-      if (complete_img(img, false)) {
-        return capture_e::error;
+      case lfa::replace_surface_with_img: {
+        auto p_surface = std::get_if<texture2d_t>(&last_frame_variant);
+        if (!p_surface) {
+          BOOST_LOG(error) << "Logical error at " << __FILE__ << ":" << __LINE__;
+          return capture_e::error;
+        }
+
+        std::shared_ptr<platf::img_t> img;
+        if (!pull_free_image_cb(img)) return capture_e::interrupted;
+
+        auto [d3d_img, lock] = get_locked_platform_img(img);
+        if (!d3d_img) return capture_e::error;
+
+        device_ctx->CopyResource(d3d_img->capture_texture.get(), p_surface->get());
+
+        last_frame_variant = img;
+        break;
       }
 
-      // We have a previously captured frame to reuse. We can't just grab the src texture from
-      // the call to AcquireNextFrame() because that won't be valid. It seems to return a texture
-      // in the unmodified desktop format (rather than the formats we passed to DuplicateOutput1())
-      // if called in that case.
-      status = img->capture_mutex->AcquireSync(0, INFINITE);
-      if (status != S_OK) {
-        BOOST_LOG(error) << "Failed to acquire capture mutex [0x"sv << util::hex(status).to_string_view() << ']';
-        return capture_e::error;
+      case lfa::replace_img_with_surface: {
+        auto p_img = std::get_if<std::shared_ptr<platf::img_t>>(&last_frame_variant);
+        if (!p_img) {
+          BOOST_LOG(error) << "Logical error at " << __FILE__ << ":" << __LINE__;
+          return capture_e::error;
+        }
+        auto [d3d_img, lock] = get_locked_platform_img(*p_img);
```

Comment:
Do we need to lock the image here? It doesn't look like we use `d3d_img` in this codepath.

As an aside, this case will always be followed by `copy_last_surface_and_blend_cursor` which calls `pull_free_image_cb()`. Is it worth including a code path that allows us to just use the `img_t` we have here instead of having to call `pull_free_image_cb()` again?

Reply:
> I'll see if I can do it without making the code significantly less readable.

One more path on `output_frame_action` switch and one temporary variable, close to no duplication. Should be alright  https://github.com/ns6089/Sunshine/commit/8ecd15794fa6181e47a33fec4598fc134f0734bd

Reply:
And now thinking about it, we can't draw to this image exactly because it may be still used by encoder(s), locks don't help because they can race. Original code was correct.

Reply:
> We don't write to it, but we read it (copy to newly allocated surface). And the lock may be held by encoder thread at this point. I don't think it's allowed by d3d, there's evidence they wanted to include read-only locks, but didn't do it in the end for some reason.

Ah yes, I didn't see that for some reason.

> And now thinking about it, we can't draw to this image exactly because it may be still used by encoder(s), locks don't help because they can race. Original code was correct. Theoretically we can check for refcount, but I'd rather not introduce not obvious strong coupling here, it screams potential problems.

Indeed, since we were handing out references to this frame, the encoder might still have one. I agree, let's just keep that logic the way you had it in the original code.